src/main/java/wellnus/gamification/util/GamificationData.java

```diff
@@ -33,13 +43,16 @@ private static int getLevelFor(int xp) {
      * Increases the user's XP points by the given amount.
      * @param pointsToAdd Number of XP points to increase
      * @return Whether the user just levelled up
+     * @throws StorageException If latest XP statistics cannot be saved to storage successfully
      */
-    public boolean addXp(int pointsToAdd) {
+    public boolean addXp(int pointsToAdd) throws StorageException {
         assert pointsToAdd > 0 : String.format(INVALID_EXP_POINTS_TO_ADD_ERROR, pointsToAdd);
         xp += pointsToAdd;
         int newLevel = getLevelFor(xp);
         boolean hasLevelledUp = newLevel > level;
         level = newLevel;
+        GamificationStorage storage = new GamificationStorage();
+        storage.store(this);

```

Comment:
Minor nit for cohesiveness, but renaming this to `gamificationStorage` might be better in case a reader/tester mistakenly thinks this is referring to `Storage` class.